src/Sarif/Visitors/RebaseUriVisitor.cs

```diff
@@ -0,0 +1,187 @@
+
+        /// <summary>
+        /// If we are changing the URIs in Results to be relative, we need to also change the URI keys in the files dictionary
+        /// to be relative.
+        /// 
+        /// For FileData, we need to fix up the URI data (making it relative to the appropriate base address), 
+        /// and also fix up the ParentKey (as we are patching the keys in the files dictionary in the Run).
+        /// (We need to fix up the keys as we are patching the PhysicalLocation in the Result objects.)
+        /// </summary>
+        /// <param name="run">A run to fix the Files dictionary of.</param>
+        internal void FixFiles(Run run)
+        {
+            Dictionary<string, FileData> newDictionary = new Dictionary<string, FileData>();
+
+            foreach (var key in run.Files.Keys)
+            {
+                Uri oldUri;
+                string newKey = key;
+                FileData data = run.Files[key];
+                // If the old uri is absolute and we need to rebase it, we should.
+                if (Uri.TryCreate(key, UriKind.Absolute, out oldUri) && oldUri.IsAbsoluteUri && _baseUri.IsBaseOf(oldUri))
+                {
+                    Uri newUri = _baseUri.MakeRelativeUri(oldUri);
+
+                    // Ensure the filedata reflects the correct base URI details.
+                    if (data != null)
+                    {
+                        if (data.ParentKey != null)
+                        {
+                            Uri parentUri;
+                            if (Uri.TryCreate(data.ParentKey, UriKind.Absolute, out parentUri))
+                            {
+                                if (oldUri.IsAbsoluteUri && _baseUri.IsBaseOf(parentUri))
+                                {
+                                    data.ParentKey = _baseUri.MakeRelativeUri(new Uri(data.ParentKey)).ToString();
+                                }
+                            }
+                        }
+
+                        data.Uri = newUri;
+                        data.UriBaseId = _baseName;
+                    }
+
+                    if(newDictionary.ContainsKey(newUri.ToString()))
+                    {
+                        throw new InvalidOperationException("Cannot rebase this file, as two URIs will collide in the file dictionary.");
+                    }
+                    newKey = newUri.ToString();
+                }
+
+                newDictionary[newKey] = data;
+            }
+
+            run.Files = newDictionary;
+        }
+
+        internal static bool TryDeserializePropertyDictionary(SerializedPropertyInfo serializedProperty, out Dictionary<string, Uri> dictionary)
+        {
+            try
+            {
+                dictionary = JsonConvert.DeserializeObject<Dictionary<string, Uri>>(serializedProperty.SerializedValue, _settings);
+
+                return true;
+            }
+            // Didn't deserialize correctly
+            catch (Exception)
```

Comment:
>Exception [](start = 19, length = 9)

please catch a more specific exception, i.e., the one that's actually generated when we hit our known serialization bug